Add specializations of multiply to reduce overpromotions to var

mcol opened the pull request:
## Summary

This is an attempt to fix #355. It adds some specializations to `rev/mat/fun/multiply.hpp` so that arithmetic types don't get promoted to `var`. I'm not entirely sure this is what was desired, but we can take it from there.

## Tests

Current tests should be enough.

## Side Effects

None.

## Checklist

- [X] Math issue #355

- [X] Copyright holder: Marco Colombo

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [X] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [X] the code is written in idiomatic C++ and changes are documented in the doxygen

- [X] the new changes are tested

SteveBronder commented on stan/math/rev/mat/fun/multiply.hpp:
```diff
@@ -598,11 +608,21 @@ template <typename T1, int R1, int C1, typename T2,
+
+template <int R1, int C1, typename T2, typename = require_arithmetic_t<T2>>
+inline Eigen::Matrix<var, R1, C1> multiply(const Eigen::Matrix<var, R1, C1>& m,
+                                           const T2& c) {
```
Did you try something like

```cpp
template <typename EigVar, typename Arith, 
   require_eigen_vt<is_var, EigVar>...,
   require_arithmetic_t<Arith>...>
inline Eigen::Matrix<var, R1, C1> multiply(EigVar&& m, Arith&& c) {
```

mcol replied:
I've tried
```
template <typename EigVar, typename Arith,
          require_eigen_vt<is_var, EigVar>..., require_arithmetic_t<Arith>...>
inline EigVar multiply(EigVar&& m, Arith&& c) {
```
and
```
template <typename EigVar, int R, int C, typename Arith,
          require_eigen_vt<is_var, EigVar>..., require_arithmetic_t<Arith>...>
inline Eigen::Matrix<var, R, C> multiply(EigVar&& m, Arith&& c) {
```
and it both cases this was not chosen over the unspecialised (preexisting) one, also if I specify `const` for the parameters.

mcol commented:
I've cleaned up a bit my previous attempt, using some of the naming suggestions proposed by @SteveBronder and added docs. I've also removed the original versions that used `to_var` by adding versions where both parameters are of var type: this passes the multipy mix tests, so I think it's ok.

mcol commented:
Thanks for merging (and probably also restarting the test)!
